### DISCUSSION TITLE
Fortls not showing errors (NeoVim)

### DISCUSSION THREAD

**Discussion by Victoriam7** [fortran-lang/fortls#195] [General]
**Describe the bug**
When using fortls no errors are shown

**To Reproduce**
Use default configuration
```
require'lspconfig'.fortls.setup{
    cmd = {
        'fortls',
        '--lowercase_intrisics',
        '--hover_signature',
        '--hover_language=fortran',
        '--use_signature_help'
    }
}
```
hello.f90
```
program hello
    ! This is a comment
    print *, 'hello'
    alsjdf;lajsdlfjsa
end program hello
```

**Expected behavior**
Some sort of error message

**Setup information (please complete the following information):**
 - OS: Linux
 - Python Version: 3.10.7
 - fortls Version: 2.13.0
 - Code editor used: NeoVim

**Comment by gnikit** [comment]
Why is this a bug? Which of the supported diagnostics are not produced?

**Comment by jatkinson1000** [comment]
I am experiencing a similar issue that I would appreciate any help with.
I have fortls installed and it is in my path.
My `init.lua` matches that above.

However, opening the following with `nvim testprog.f90` produces no errors.

```fortran
program one
  implicit none

  intger :: i
  real :: i

  do i = 1, 10
    write(*,*) i
  nd do

end program
```

`nvim-lspconfig` appears to be working, as a similar test on some faulty python code with pyright produces warnings at the end of lines.

I suspect, therefore, that it's an issue with how fortls is running or interfacing with neovim.
I have checked and fortls is in my path, and is being recognised by Neovim.

I note that running LSPInfo produces the below result:
<img width="1442" alt="image" src="https://github.com/fortran-lang/fortls/assets/109271713/2e20c899-b596-4799-a64d-e25139c78d85">

The log report is as follows:
```
[START][2023-10-07 18:03:27] LSP logging initiated
[ERROR][2023-10-07 18:03:27] .../vim/lsp/rpc.lua:734	"rpc"	"/opt/homebrew/bin/fortls"	"stderr"	"[INFO - 18:03:27] fortls - Fortran Language Server 2.13.0 Initialized\n"
[ERROR][2023-10-07 18:03:27] .../vim/lsp/rpc.lua:734	"rpc"	"/opt/homebrew/bin/fortls"	"stderr"	"[INFO - 18:03:27] fortls initialization complete\n"
```

Any assistance or insight would be much appreciated.

**Reply by jatkinson1000** [reply]
I should add that I get the same issue on both macOS and Linux (Arch).
I have also just installed and checked lua_ls which is also working as expected (showing errors at the EoL for faulty code).

**Reply by Victoriam7** [reply]
It's not finding a root directory. I'm pretty sure that can sometimes cause an issue. You could try making an empty git repository in the root directory and see if that fixes it.

**Reply by jatkinson1000** [reply]
Thanks.
I did that and it now shows the root directory as you would expect, but still no errors being raised for invalid lines of code.

**Reply by gnikit** [reply]
Fortls is *not* a compiler nor a linter and for that reason it offers very limited diagnostics. Please check the documentation webpage for the available diagnostics of the language server.

**Reply by jatkinson1000** [reply]
Thank you.
I am following your documentation [here](https://fortls.fortran-lang.org/editor_integration.html#neovim).
I would expect some of the features listed [here](https://fortls.fortran-lang.org/features.html) to appear once fortls is running, as they do for pyright or lua_ls.
Looking at the diagnostics listed [here](https://fortls.fortran-lang.org/features.html#diagnostics) I am not seeing any report for multiple variable definitions, unclosed code blocks, or unknown use statements.
Is there any way I can see if fortls is working behind the scenes so I can narrow down where the issue is?
I tried running via command line using `fortls --debug_filepath myfile.f90 --debug_rootpath ./ --debug_log` and got the following output:
```
DEBUG:fortls.constants:REQUEST None None
INFO:fortls.constants:fortls - Fortran Language Server 2.13.0 Initialized
```
based on the below code:
```
program one
  implicit none

  integer :: i
  real :: i

  do i = 1, 10
    write(*,*) i

end program
```

If I run with `--debug_parser` as well then I get the following output to terminal:
```
Testing parser
  File = "fcode.f90"
  Detected format: free

=========
Parser Output
=========

=== No PreProc ===

program one !!! PROGRAM - Ln:1
  implicit none !!! IMPLICIT - Ln:2
  integer :: i !!! VARIABLE - Ln:4
  real :: i !!! VARIABLE - Ln:5
  do i = 1, 10 !!! DO - Ln:7

=========
Object Tree
=========

1: one
  6: one::i
  6: one::i
  11: one::#do1
11: one::#do1

=========
Exportable Objects
=========

1: one
```

**Reply by gnikit** [reply]
From the parser outputs you posted it appears that fortls is working as expected.
The root directory should be set automatically via neovim or lspconfig, what version of neovim are you using?
When you open a Fortran file in neovim, can you check the running processes and see if fortls is running?

The issue here is the configuration, not the server itself.

Also AFAIK the neovim configuration is stored in `~/.config/nvim/init.vim`, the config options listed are in Vimscript, not sure if neovim can run Vimscript from a `.lua` file.

**Reply by jatkinson1000** [reply]
Hi there, thanks for looking at this.
I'm on Neovim version 0.9.4

It's possible to execute vimscript via lua using `vim.call()`, but I think [the instructions on the fortls website](https://fortls.fortran-lang.org/editor_integration.html#neovim) are already written for the `init.lua` setup.
I have
```
lspconfig.fortls.setup{cmd={"fortls", "--notify_init", "--hover_signature", "--hover_language=fortran", "--use_signature_help"}}
```
or
```
lspconfig.fortls.setup{}
```

I had a play around with a few things and it now seems to be working.
I'm not 100% sure what tipped it into action I'm afraid, but one thing that I can say for sure is that you need the file to be in a git repo as @Victoriam7 suggested above. Is this a general requirement of fortls?

This is on linux, I'll try OSX tomorrow if I can, and will let you know if I have any more issues as it's now set up.

Thanks again.

**Reply by gnikit** [reply]
Well when I wrote the docs, I only used `init.vim`. I was never a heavy neovim user, just a vim one, so I wrote the instructions based on what I could figure out back then. This is the relevant bit of my config from then:

```lua
-- Use a loop to conveniently call 'setup' on multiple servers and
-- map buffer local keybindings when the language server attaches
local servers = { 'pyright', 'rust_analyzer', 'tsserver'}
for _, lsp in pairs(servers) do
  require('lspconfig')[lsp].setup {
    on_attach = on_attach,
    flags = {
      -- This will be the default in neovim 0.7+
      debounce_text_changes = 150,
    }
  }
end

require'lspconfig'.fortls.setup{
    on_attach = on_attach,
    cmd = {
      'fortls',
      '--notify_init',
      '--hover_signature',
      '--hover_language=fortran',
      '--use_signature_help',
    },
}
```

> I'm not 100% sure what tipped it into action I'm afraid, but one thing that I can say for sure is that you need the file to be in a git repo as @Victoriam7 suggested above. Is this a general requirement of fortls?

No, fortls works both for directories (with and without git) and individual files. The VS Code extension showcases both.

**Comment by abzrgzdh** [comment]
Sorry to continue this somewhat old discussion. I also have this problem. The autocompletion works, thus the process is running, but when I introduce an error in the code, no diagnostics are emitted. Usually existing errors are populated in the quick fix list but none are in that list. Here's my config:

```lua
    local lspconfig = require('lspconfig')
-- ...
    lspconfig.fortls.setup {
      cmd = {
        "fortls",
        "--nthreads", "2",
        "--symbol_skip_mem",
        "--autocomplete_no_prefix",
        "--incremental_sync",
        "--use_signature_help",
        "--variable_hover",
        "--hover_signature",
        "--enable_code_actions",
        "--lowercase_intrinsics",
      },
      filetypes = { "fortran" },
      root_dir = require('lspconfig.util').root_pattern(
        '.fortls', '.git'
      ) or function() return vim.loop.cwd() end,
      capabilities = capabilities
    }
```

In vscode as soon as I save the code it will emit the errors. Not sure if I'm missing some configuration or additional programs that is included in vscode but not in my setup.

- OS: macos
- Neovim:
  ```
  NVIM v0.10.0
  Build type: Release
  LuaJIT 2.1.1713484068
  Run "nvim -V1 -v" for more info
  ```
- Fortls: 3.1.2
